File: elementary_data_structures/linked_lists.py
class Element:

    __slots__ = ['data', 'next', 'prev']

    def __init__(self, data):
        self.data = data
        self.next = self.prev = None


class LinkedList:

    __slots__ = ['size', 'head', 'tail']

    def __init__(self):
        self.size = 0
        self.head = self.tail = None

    def append(self, z):
        self.size += 1

        if not self.head:
            self.head = z
        else:
            self.tail.next = z
        z.prev = self.tail
        self.tail = z

    def contains(self, data):
        p = self.head
        while p and p.data != data:
            p = p.next

        return p

    def remove(self, z):
        self.size -= 1

        if z.next:
            z.next.prev = z.prev
        else:
            self.tail = z.prev
        if z.prev:
            z.prev.next = z.next
        else:
            self.head = z.next
        z = None

    def print_list(self):
        print(f"size = {self.size} : ")

        print("--- from head to tail ---")
        p = self.head
        while p:
            print(f"{p.data}", end=' -> ' if p.next else '\n')

            p = p.next

        print("--- from tail to head ---")
        p = self.tail
        while p:
            print(f"{p.data}", end=' <- ' if p.prev else '\n')

            p = p.prev


if __name__ == '__main__':
    ll = LinkedList()

    while True:
        print("1:append 2:remove 3:print > ", end='')
        op = int(input())

        if op == 1:
            print("input data > ", end='')
            data = int(input())

            ll.append(Element(data))
        elif op == 2:
            print("input data > ", end='')
            data = int(input())

            z = ll.contains(data)
            if not z:
                print(f"{data} is not found in the list.")
            else:
                ll.remove(z)
        elif op == 3:
            ll.print_list()
        else:
            print("invalid operation")

File: elementary_data_structures/linked_lists.go
//go:build ignore

package main

import "fmt"

type Element struct {
	data int
	next *Element
	prev *Element
}

type List struct {
	size int
	head *Element
	tail *Element
}

func append(l *List, z *Element) {
	l.size += 1

	if l.head == nil {
		l.head = z
	} else {
		l.tail.next = z
	}
	z.prev = l.tail
	l.tail = z
}

func contains(l *List, data int) *Element {
	for p := l.head; p != nil; p = p.next {
		if p.data == data {
			return p
		}
	}

	return nil
}

func remove(l *List, z *Element) {
	l.size -= 1

	if z.next != nil {
		z.next.prev = z.prev
	} else {
		l.tail = z.prev
	}
	if z.prev != nil {
		z.prev.next = z.next
	} else {
		l.head = z.next
	}
	z = nil
}

func printList(l *List) {
	fmt.Printf("size = %d\n", l.size)

	fmt.Println("--- from head to tail ---")
	for p := l.head; p != nil; p = p.next {
		fmt.Printf("%d", p.data)
		if p.next != nil {
			fmt.Print(" -> ")
		}
	}
	fmt.Println()

	fmt.Println("--- from tail to head ---")
	for p := l.tail; p != nil; p = p.prev {
		fmt.Printf("%d", p.data)
		if p.prev != nil {
			fmt.Print(" <- ")
		}
	}
	fmt.Println()
}

func main() {
	l := List{0, nil, nil}

	for {
		fmt.Print("1:append 2:remove 3:print > ")
		var op int
		fmt.Scanf("%d", &op)

		switch op {
		case 1:
			fmt.Print("input data > ")
			var data int
			fmt.Scanf("%d", &data)

			z := Element{data, nil, nil}
			append(&l, &z)
		case 2:
			fmt.Print("input data > ")
			var data int
			fmt.Scanf("%d", &data)

			z := contains(&l, data)
			if z == nil {
				fmt.Printf("%d is not found in the list.\n", data)
			} else {
				remove(&l, z)
			}
		case 3:
			printList(&l)
		default:
			fmt.Println("invalid operation")
		}
	}
}
